syncer/optimist.go

```diff
@@ -0,0 +1,188 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package syncer
+
+import (
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/tidb-tools/pkg/filter"
+	"github.com/siddontang/go-mysql/replication"
+	"go.uber.org/zap"
+
+	"github.com/pingcap/dm/pkg/log"
+	"github.com/pingcap/dm/pkg/shardddl/optimism"
+	"github.com/pingcap/dm/pkg/terror"
+)
+
+// trackedDDL keeps data needed for schema tracker.
+type trackedDDL struct {
+	rawSQL     string
+	stmt       ast.StmtNode
+	tableNames [][]*filter.Table
+}
+
+// initOptimisticShardDDL initializes the shard DDL support in the optimistic mode.
+func (s *Syncer) initOptimisticShardDDL() error {
+	// fetch tables from source and filter them
+	sourceTables, err := s.fromDB.fetchAllDoTables(s.bwList)
+	if err != nil {
+		return err
+	}
+
+	return s.optimist.Init(sourceTables)
+}
+
+// handleQueryEventOptimistic handles QueryEvent in the optimistic shard DDL mode.
+func (s *Syncer) handleQueryEventOptimistic(
+	ev *replication.QueryEvent, ec eventContext,
+	needHandleDDLs []string, needTrackDDLs []trackedDDL,
+	onlineDDLTableNames map[string]*filter.Table) error {
+	// wait previous DMLs to be replicated
+	s.jobWg.Wait()
+
+	var (
+		upSchema   string
+		upTable    string
+		downSchema string
+		downTable  string
+	)
+	for _, td := range needTrackDDLs {
+		// check whether do shard DDL for multi upstream tables.
+		if upSchema != "" && upSchema != td.tableNames[0][0].Schema &&
+			upTable != "" && upTable != td.tableNames[0][0].Name {
+			return terror.ErrSyncerUnitDDLOnMultipleTable.Generate(string(ev.Query))
+		}
+		upSchema = td.tableNames[0][0].Schema
+		upTable = td.tableNames[0][0].Name
+		downSchema = td.tableNames[1][0].Schema
+		downTable = td.tableNames[1][0].Name
+	}
+
+	var tiBefore *model.TableInfo
+	if _, ok := needTrackDDLs[0].stmt.(*ast.CreateTableStmt); !ok {
+		var err error
+		tiBefore, err = s.getTable(upSchema, upTable, downSchema, downTable, ec.parser2)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, td := range needTrackDDLs {
+		if err := s.trackDDL(string(ev.Schema), td.rawSQL, td.tableNames, td.stmt, &ec); err != nil {
+			return err
+		}
+	}
+
+	// TODO(csuzhangxc): rollback schema in the tracker if failed.
+	tiAfter, err := s.getTable(upSchema, upTable, downSchema, downTable, ec.parser2)
+	if err != nil {
+		return err
+	}
+
+	s.tctx.L().Info("save table checkpoint", zap.String("event", "query"),
+		zap.String("schema", upSchema), zap.String("table", upTable),
+		zap.Strings("ddls", needHandleDDLs), log.WrapStringerField("location", ec.currentLocation))
+	s.saveTablePoint(upSchema, upTable, ec.currentLocation.Clone())
```

Comment:
why save table point here? do we need to save it after ddl is executed?

Reply:
Now, it's useless, but If we need to support _handle some types of conflicts automatically_ , then we need _redirect streamer_ as we did for the pessimistic shard mode (then `s.checkpoint.IsNewerTablePoint` will be needed), so I add this just like pessimistic shard mode did now.